substrate/primitives/io/src/global_alloc_wasm.rs

```diff
@@ -17,69 +17,177 @@
+
+// SAFETY: This is runtime-only, and runtimes are single-threaded, so this is safe.
+unsafe impl Send for LocalAllocator {}
+
+// SAFETY: This is runtime-only, and runtimes are single-threaded, so this is safe.
+unsafe impl Sync for LocalAllocator {}
+
+static LOCAL_ALLOCATOR: LocalAllocator =
+	LocalAllocator(UnsafeCell::new(picoalloc::Allocator::new(RuntimeAllocator)));
+
+fn local_allocator() -> &'static mut picoalloc::Allocator<RuntimeAllocator> {
+	// SAFETY: This is only called when allocating memory, and the allocator
+	// doesn't trigger the global allocator recursively, so only a single
+	// &mut will ever exist at the same time.
+	unsafe { &mut *LOCAL_ALLOCATOR.0.get() }
+}
+
+/// Checks whether a given pointer came from the local allocator.
+fn is_local_pointer(ptr: *mut u8) -> bool {
+	ptr.addr() >= LOCAL_HEAP.0.get().addr() &&
+		ptr.addr() < LOCAL_HEAP.0.get().addr() + LOCAL_HEAP_SIZE
+}
+
 unsafe impl GlobalAlloc for RuntimeAllocator {
 	unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
-		let align = layout.align();
-		let size = layout.size();
+		// These should never fail, but let's do proper error checking anyway.
+		let Some(align) = picoalloc::Size::from_bytes_usize(layout.align()) else {
+			return core::ptr::null_mut();
+		};
+
+		let Some(size) = picoalloc::Size::from_bytes_usize(layout.size()) else {
+			return core::ptr::null_mut();
+		};
+
+		// Try to allocate memory from the local pool, and only fall back
+		// to the host allocator if this fails.
+		if let Some(pointer) = local_allocator().alloc(align, size) {
+			pointer.as_ptr()
+		} else {
+			crate::global_alloc_wasm_legacy::HostAllocator.alloc(layout)
+		}
+	}
 
-		// Allocate for the required size, plus a potential alignment.
-		//
-		// As the host side already aligns the pointer by `8`, we only need to account for any
-		// excess.
-		let ptr = crate::allocator::malloc((size + align.saturating_sub(8)) as u32);
+	unsafe fn dealloc(&self, ptr: *mut u8, _: Layout) {
+		if is_local_pointer(ptr) {
+			// SAFETY: We've checked that the pointer is from the local allocator.
+			unsafe { local_allocator().free(NonNull::new_unchecked(ptr)) }
+		} else {
+			crate::global_alloc_wasm_legacy::HostAllocator.dealloc(ptr)
+		}
+	}
 
-		// Calculate the required alignment.
-		let ptr_offset = ptr.align_offset(align);
+	unsafe fn alloc_zeroed(&self, layout: Layout) -> *mut u8 {
+		let Some(align) = picoalloc::Size::from_bytes_usize(layout.align()) else {
+			return core::ptr::null_mut();
+		};
 
-		// Should never happen, but just to be sure.
-		if ptr_offset > u16::MAX as usize || ptr.is_null() {
-			return ptr::null_mut()
+		let Some(size) = picoalloc::Size::from_bytes_usize(layout.size()) else {
+			return core::ptr::null_mut();
+		};
+
+		// First try the local allocator. Use its `alloc_zeroed` as its
+		// smart enough to not unnecessarily zero-fill the memory if it's
+		// the very first allocation which touches this region of the heap.
```

Comment:
dq: how it is ensured that it is the first allocation that touches the region of the heap?

Reply:
https://github.com/koute/picoalloc/blob/a314b2aa06bca69c7518c465117d5925fc156dbe/src/allocator.rs#L827